rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,261 @@
+pass it to the auth server.
+
+### EC2 Method
+
+In place of a join token, nodes will present an EC2 IID (Instance Identity
+Document) to the auth server, with a corresponding PKCS7 signature.
+
+This will be sent along with the other existing parameters in a
+[RegisterUsingTokenRequest](
+https://github.com/gravitational/teleport/blob/d4247cb150d720be97521347b74bf9c526ae869f/lib/auth/auth.go#L1538-L1563).
+
+The auth server will then:
+1. Use the AWS DSA public certificate to check that the PKCS7 signature for the
+   IID is valid (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-pkcs7.html).
+2. Check that the `pendingTime` of the IID, which is the time the instance was
+   launched, is within a 5 minute TTL.
+   - if the node fails to join the cluster during this window, the user can
+     stop and restart the EC2 instance to reset the `pendingTime` (and
+     effectively create a new IID with a new signature)
+3. Check that the AWS join token matches the AWS account in the IID, and the
+   requested Teleport service role.
+4. Check that this EC2 instance has not already joined the cluster.
+   - The node name will be set to `<aws_account_id>-<aws_instance_id>` so that
+     the auth server can efficiently check the backend to see whether this
+     instance has already joined.
+   - This is to prevent replay attacks with the IID
+5. Check that the EC2 instance is currently running. \*
+
+\* Step 5 requires AWS credentials on the Auth server with permissions for
+`ec2:DescribeInstances`. If you need Nodes to be able to join the teleport
+cluster from different AWS accounts than the one in which the Teleport Auth
+server is running, there are a few extra required steps:
+1. Every AWS account in which Nodes should be able to join will need an IAM
+   role which can be assumed from the account of the Auth server. This role
+   needs permissions for `ec2:DescribeInstances`.
+2. The IAM role of the Auth server needs permissions to `sts:AssumeRole` for
+   the ARN of every foreign IAM role from step 1. These ARNs will need to be
+   configured in the AWS join token.
+
+### IAM Method
+
+In place of a join token, nodes will present a signed `sts:GetCallerIdentity`
+request to the auth server.
+
+The form of this request is an HTTP `POST` request to `sts.amazonaws.com` with
+`Action=GetCallerIdentity` in the body. The request can include arbitrary
+headers which will also be signed using the
+[Signature Version 4 signing process](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html).
+The signature will then be included in the `Authorization` header of the request.
+See Appendix I for an example.
+
+The actual creation and signing of the request will be done on the node with
+the AWS SDK for Go.
+
+Because the signed request can include arbitrary headers, this allows us to
+issue a challenge (a crypto random string of bytes) that the node must include
```

Comment:
The lifetime of this challenge should be limited.

Reply:
add a 1 minute timeout to the doc